modules/ocl/cl_kernel.cpp

```diff
@@ -111,19 +119,75 @@ CLKernel::build_kernel (const XCamKernelInfo& info, const char* options)
+        if (cache_path != NULL) {
+            if (access (cache_path, F_OK) == -1) {
+                mkdir (cache_path, S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);
+            }
+            memset (_kernel_cache_path, 0, XCAM_MAX_STR_SIZE);
+            snprintf (
+                _kernel_cache_path, XCAM_MAX_STR_SIZE - 1,
+                "%s/%s",
+                cache_path, key_str);
+        } else {
+            if (access (_kernel_cache_path, F_OK) == -1) {
+                mkdir (_kernel_cache_path, S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);
+            }
+            strncat (_kernel_cache_path, key_str, XCAM_MAX_STR_SIZE - 1);
+        }
+
         i_kernel = _kernel_map.find (key);
         if (i_kernel == _kernel_map.end ()) {
             SmartPtr<CLContext>  context = get_context ();
             single_kernel = new CLKernel (context, info.kernel_name);
             XCAM_ASSERT (single_kernel.ptr ());
-            ret = single_kernel->load_from_source (info.kernel_body, strlen (info.kernel_body), NULL, NULL, options);
-            XCAM_FAIL_RETURN (
-                ERROR, ret == XCAM_RETURN_NO_ERROR, ret,
-                "build kernel(%s) from source failed", key_str);
-            //kernel_map.insert (std::make_pair (key, single_kernel));
-            _kernel_map[key] = single_kernel;
-        } else
+
+            FileHandle cache_file;
+            size_t cache_size = 0;
+            uint8_t *kernel_cache = NULL;
+            bool load_cache = false;
+
+            ret = cache_file.open (_kernel_cache_path, "r");
+            if (ret == XCAM_RETURN_NO_ERROR) {
+                cache_file.get_file_size (cache_size);
+                if (cache_size > 0) {
+                    kernel_cache = (uint8_t*) xcam_malloc0 (sizeof (uint8_t) * (cache_size + 1));
+                    cache_file.read_file (kernel_cache, cache_size);
+                    cache_file.close ();
+
+                    ret = single_kernel->load_from_binary (kernel_cache, cache_size);
+                    XCAM_FAIL_RETURN (
+                        ERROR, ret == XCAM_RETURN_NO_ERROR, ret,
+                        "build kernel(%s) from binary failed", key_str);
+
+                    load_cache = true;
+                    xcam_free (kernel_cache);
+                }
+            } else {
+                XCAM_LOG_DEBUG ("open kernel cache file to read failed ret(%d)", ret);
+            }
+
+            if (load_cache == false) {
+                ret = single_kernel->load_from_source (info.kernel_body, strlen (info.kernel_body), &kernel_cache, &cache_size, options);
+                XCAM_FAIL_RETURN (
+                    ERROR, ret == XCAM_RETURN_NO_ERROR, ret,
+                    "build kernel(%s) from source failed", key_str);
+                //kernel_map.insert (std::make_pair (key, single_kernel));
+                _kernel_map[key] = single_kernel;
+
+                ret = cache_file.open (_kernel_cache_path, "wb");
```

Comment:
You should write the binary to a temporal unique file(e.g. dot+kernel_name+keyid+milli-second-timestamp) in case other process is reading the final file. After the file done successfully , then rename  the file to final file name.  and also need check file.write() return value in writing. if any error happened, then delete this tmp file.